game/layout/layoutloader.py

```diff
@@ -0,0 +1,203 @@
+            for file in Path(TEMPLATE_DIR).rglob("*.yaml"):
+                if not file.is_file():
+                    continue
+                with file.open("r", encoding="utf-8") as f:
+                    mapping_dict = yaml.safe_load(f)
+
+                template_map = LayoutMapping.from_dict(mapping_dict, f.name)
+
+                if template_map.layout_file in mappings:
+                    mappings[template_map.layout_file].append(template_map)
+                else:
+                    mappings[template_map.layout_file] = [template_map]
+
+        with logged_duration(f"Parsing all layout miz multithreaded"):
+            with ThreadPoolExecutor() as exe:
+                for miz, maps in mappings.items():
+                    exe.submit(self._load_from_miz, miz, maps)
+
+        logging.info(f"Imported {len(self._templates)} layouts")
+        self._dump_templates()
+
+    def _dump_templates(self) -> None:
+        file = Path(persistency.base_path()) / TEMPLATE_DUMP
+        dump = (VERSION, self._templates)
+        with file.open("wb") as fdata:
+            pickle.dump(dump, fdata)
+
+    @staticmethod
+    def mapping_for_group(
+        mappings: list[LayoutMapping], group_name: str
+    ) -> tuple[LayoutMapping, int, GroupLayoutMapping]:
+        for mapping in mappings:
+            for g_id, group_mapping in enumerate(mapping.groups):
+                if (
+                    group_mapping.name == group_name
+                    or group_name in group_mapping.statics
+                ):
+                    return mapping, g_id, group_mapping
+        raise KeyError
+
+    def _load_from_miz(self, miz: str, mappings: list[LayoutMapping]) -> None:
+        template_position: dict[str, Point] = {}
+        temp_mis = dcs.Mission()
+        with logged_duration(f"Parsing {miz}"):
+            # The load_file takes a lot of time to compute. That's why the layouts
+            # are written to a pickle and can be reloaded from the ui
+            # Example the whole routine: 0:00:00.934417,
+            # the .load_file() method: 0:00:00.920409
+            temp_mis.load_file(miz)
+
+        for country in itertools.chain(
+            temp_mis.coalition["red"].countries.values(),
+            temp_mis.coalition["blue"].countries.values(),
+        ):
+            for dcs_group in itertools.chain(
+                temp_mis.country(country.name).vehicle_group,
+                temp_mis.country(country.name).ship_group,
+                temp_mis.country(country.name).static_group,
+            ):
+                try:
+                    mapping, group_id, group_mapping = self.mapping_for_group(
+                        mappings, dcs_group.name
+                    )
+                except KeyError:
+                    logging.warning(f"No mapping for dcs group {dcs_group.name}")
+                    continue
+
+                template = self._templates.get(mapping.name, None)
+                if template is None:
```

Comment:
The reason for this is that within a miz can be more than 1 layout. Also multiple dcs groups are part of a layout but it can not be ensured that all groups of one layout come in the correct order. It is possible that dcs group 1 is part of layout A and dcs group 2 of layout B. Thats why we need to read the layout by name and if it exists add the group to the layout. Otherwise create a new layout. I do not see another way to solve this problem without defining restrictions for the miz file which is only possible once all layouts are migrated / generalized.